File: 2018-02/2018-02-22/leetcode123.java
/**
 * date: 2018-02-22
 * author: pwxcoo
 * complexity: T = O(n), S = O(1)
 * describe:    dp。太屌了这个。
 */

class Solution {
    public int maxProfit(int[] prices) {
        int hold1 = Integer.MIN_VALUE, hold2 = Integer.MIN_VALUE;
        int pro1 = 0, pro2 = 0;
        
        for(int i : prices)
        {
            pro2 = Math.max(pro2, hold2 + i);
            hold2 = Math.max(hold2, pro1 - i);
            pro1 = Math.max(pro1, hold1 + i);
            hold1 = Math.max(hold1, -i);
        }
        
        return pro2;
    }
}

File: 2018-08/2018-08-06/leetcode123.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int len = prices.size();
        if (len == 0) return 0;
        
        int T = 2;
        vector<vector<int>> dp(T + 1, vector<int>(len, 0));
        for (int t = 1; t <= 2; t++)
        {
            int min_before = prices[0];
            for (int i = 1; i < len; i++)
            {
                dp[t][i] = max(dp[t][i - 1], prices[i] - min_before);
                min_before = min(min_before, prices[i] - dp[t - 1][i - 1]);
            }
        }
        
        return dp[2][len - 1];
    }
};